zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -1003,6 +1004,32 @@ public static List<ACL> fixupACL(String path, List<Id> authInfo, List<ACL> acls)
             throw new KeeperException.InvalidACLException(path);
         }
         List<ACL> rv = new ArrayList<>();
+
+        // Overwrite the acl list for users (except super user) when znode group acl feature is on
+        // Set znode ACL to the corresponding znode group acl designated ACL instead
+        // For single domain user / cross domain components -> (x509 : domainName)
+        // For user w/o a domain -> (x509: clientURI)
+        if (X509AuthenticationConfig.getInstance().isX509ClientIdAsAclEnabled() && ProviderRegistry
+            .getServerProvider(
+                X509AuthenticationUtil.X509_SCHEME) instanceof X509ZNodeGroupAclProvider
+            && !X509AuthenticationConfig.getInstance().isDedicatedServerEnabled()) {
+            authInfo.forEach(id -> {
```

Comment:
Do we expect to have multiple IDs in `authInfo`? If so, we are clearing `uniqacls` for each of them, and rewriting the list with the next ID, effectively keeping only the last one. If this is the desired behavior, we can probably avoid the `forEach`.

Reply:
Or not actually avoid it, because there are some checks involved, but regardless, do we want to keep only one of the IDs in the list?

Reply:
We allow multiple domains, and therefore, we should not overwrite every time. @mgao0 please come up with a new solution here.